File: host.py
"""The host: whatever actually performs an effect.

P-Q9 has been open since the first session — "what is the implementation
host". Measuring it rather than arguing it produced a smaller question than
expected.

Almost nothing in Planes is host-shaped. The parser reads a foreign target
as an opaque string. The analyser uses it as a label and never parses it.
Only the interpreter interprets it, in one place, by splitting on a dot and
calling `importlib`. Everything else — exact rationals, JSON at the
boundary, HTTP, module files — is implementation, not language.

So the host question is not "which language should Planes be written in".
It is "what does a host have to provide", and the answer is this file: five
capabilities and a way to resolve a foreign name.

Naming that surface is the useful part. A second host is then a piece of
work with a known size, rather than a rewrite, and the *language* stops
quietly accumulating assumptions about the machine underneath it.
"""
import json
import urllib.request


class HostError(Exception):
    """The host could not do what was asked. Distinct from a program error:
    this is the machine failing, not the program being wrong."""


class Host:
    """What a host must provide.

    Five capabilities, matching the closed effect vocabulary, plus foreign
    resolution. A host that implements these runs Planes; the language does
    not otherwise care what it is written in.

    The vocabulary is what makes this small. Because effects were closed
    early — `ask`, `read`, `write`, `show`, `clock`, `random`, `env` — the
    host surface could not sprawl. That was decided for the analyser, and it pays
    out here.
    """

    name = "abstract"

    # ---- effects

    def ask(self, url):
        """A request expecting a response. Returns the body as text."""
        raise NotImplementedError

    def read(self, path):
        """Read a file. Returns its contents as text."""
        raise NotImplementedError

    def write(self, path, text):
        """Write a file."""
        raise NotImplementedError

    def show(self, text):
        """Emit a line to wherever output goes."""
        raise NotImplementedError

    def clock(self):
        """Seconds since the epoch, as a float."""
        raise NotImplementedError

    # ---- the record plane (§99) — a host capability, not a program effect

    def record(self, entry):
        """Persist a record entry, if this host keeps one.

        Optional, unlike the five capabilities above: a host that does
        nothing here is still a complete host, and the interpreter must
        not depend on this happening. The default is a no-op.
        """

    # ---- foreign resolution

    def resolve(self, target):
        """Turn a foreign target string into something callable.

        The target is opaque to the language. A Python host reads
        `builtins.sorted`; a JavaScript host could read `node:fs#readFile`;
        a Rust host, a crate path. The *string* is host-specific by design,
        which is why no syntax change is needed to move hosts.
        """
        raise NotImplementedError

    def target_hint(self):
        """How this host wants a foreign target written, for error messages."""
        return "a name this host understands"

    # ---- data at the boundary

    def parse_json(self, text):
        raise NotImplementedError



class PythonHost(Host):
    """The host Planes runs on today.

    Chosen originally by default rather than by decision, and the decision
    is now recorded: it stays, because every requirement the language has
    accumulated is met here and the alternatives cost more than they return
    at this stage. See REPORT_HOST.md.
    """

    name = "python"

    def __init__(self):
        self._resolved = {}

    def ask(self, url):
        req = urllib.request.Request(
            url, headers={"User-Agent": "planes/0.1"})
        with urllib.request.urlopen(req, timeout=20) as r:
            return r.read().decode()

    def read(self, path):
        with open(path) as f:
            return f.read()

    def write(self, path, text):
        with open(path, "w") as f:
            f.write(text)

    def show(self, text):
        print(text)

    def clock(self):
        import time
        return time.time()

    def resolve(self, target):
        import importlib
        if target in self._resolved:
            return self._resolved[target]
        mod_path, _, attr = target.rpartition(".")
        if not mod_path:
            raise HostError(f"bad target: {target}")
        try:
            fn = getattr(importlib.import_module(mod_path), attr)
        except (ImportError, AttributeError):
            raise HostError(f"cannot find '{target}'")
        self._resolved[target] = fn
        return fn

    def target_hint(self):
        return "`module.function`, e.g. `builtins.sorted`"

    def parse_json(self, text):
        return json.loads(text)



class TestHost(PythonHost):
    """A host with the outside world replaced.

    Not a mock bolted onto tests — a host, implementing the same five
    capabilities. That it can exist at all is the evidence that the seam is
    real rather than decorative, and it makes the whole test suite hermetic
    by construction instead of by discipline.
    """

    name = "test"

    def __init__(self, responses=None, files=None, now=None):
        super().__init__()
        self.responses = responses or {}
        self.files = dict(files or {})
        self.now = now if now is not None else 1_000_000.0
        self.shown = []
        self.recorded = []      # the record plane's in-memory sink

    def ask(self, url):
        r = self.responses
        if callable(r):
            return r(url)
        if url in r:
            return r[url]
        raise HostError(f"no stubbed response for {url}")

    def read(self, path):
        if path not in self.files:
            raise HostError(f"no such file: {path}")
        return self.files[path]

    def write(self, path, text):
        self.files[path] = text

    def show(self, text):
        self.shown.append(text)

    def clock(self):
        return self.now

    def record(self, entry):
        self.recorded.append(entry)

File: js/host.mjs
// js/host.mjs — the host: whatever actually performs an effect (browser-safe).
//
// The JavaScript counterpart of host.py's abstract surface plus the in-memory
// hosts. This module imports nothing node-specific, so it — and everything that
// imports it, including interp.mjs — loads unchanged in a browser tab. The
// real-filesystem NodeHost lives in host_node.mjs; the browser VFS host in
// host_browser.mjs.
//
// host.py names seven required methods (each raises NotImplementedError
// there): ask, read, write, show, clock, resolve, parseJson. `record` is an
// optional no-op, not one of the seven.
//
// It was eight until C4 counted what the reference actually calls rather than
// what host.py declares. `toJson` had no caller anywhere: `interp.py`'s write
// effect uses a *module-level* to_json that unwraps the value model first, and
// the host method took an already-plain value, so it could not have served the
// one site that serialises even if something had reached for it.
// `test_host.py` checked the surface with `hasattr`, which is a declaration
// check and cannot tell a dead method from a live one.
//
// The seam claim is restated at seven, not spent: seven methods, not a
// rewrite. See REPORT_SECOND_HOST.md's seam verdict and REPORT_FAST_FOLLOW.md.

// The host could not do what was asked. Distinct from a program error.
export class HostError extends Error {
  constructor(message) {
    super(message);
    this.name = "HostError";
  }
}

// What a host must provide: five effect capabilities, foreign resolution, and a
// JSON boundary. A host that implements these runs Planes.
export class Host {
  get name() {
    return "abstract";
  }
  ask(_url) {
    throw new HostError("ask: not implemented on the abstract host");
  }
  read(_path) {
    throw new HostError("read: not implemented on the abstract host");
  }
  write(_path, _text) {
    throw new HostError("write: not implemented on the abstract host");
  }
  show(_text) {
    throw new HostError("show: not implemented on the abstract host");
  }
  clock() {
    throw new HostError("clock: not implemented on the abstract host");
  }
  // The optional record plane — a no-op by default; a host that does nothing
  // here is still complete.
  record(_entry) {}
  resolve(_target) {
    throw new HostError("resolve: not implemented on the abstract host");
  }
  targetHint() {
    return "a name this host understands";
  }
  parseJson(_text) {
    throw new HostError("parseJson: not implemented on the abstract host");
  }
}

// json.dumps(..., indent=2) escapes non-ASCII (ensure_ascii=True); JSON
// .stringify does not. Escape every UTF-16 unit above 0x7e as \uXXXX so a
// written file is byte-identical to interp.py's; an astral char becomes a
// surrogate pair on both sides.
export function ensureAscii(s) {
  let out = "";
  for (let i = 0; i < s.length; i++) {
    const c = s.charCodeAt(i);
    if (c > 0x7e) out += "\\u" + c.toString(16).padStart(4, "0");
    else out += s[i];
  }
  return out;
}

// json.dumps(value, indent=2) with ensure_ascii — the write effect's
// serialisation, and the only one there is. `interp.mjs`'s module-level
// `toJson` unwraps a Planes value and hands the result here; `interp.py`'s
// does the same with `json.dumps(..., indent=2)`. C4: a `toJson` also sat on
// the host surface, wrapping this, and nothing ever called it.
export function pyJsonDumps(value) {
  return ensureAscii(JSON.stringify(value, null, 2));
}

// Python's sorted/max/min ordering over a homogeneous list — foreign.planes
// resolves builtins.sorted/max/min; these reproduce the default order so a
// resolved foreign agrees with interp.py's.
export function pyCompare(a, b) {
  if (typeof a === "number" && typeof b === "number") return a - b;
  if (typeof a === "string" && typeof b === "string") {
    const as = [...a];
    const bs = [...b];
    const n = Math.min(as.length, bs.length);
    for (let i = 0; i < n; i++) {
      const d = as[i].codePointAt(0) - bs[i].codePointAt(0);
      if (d !== 0) return d;
    }
    return as.length - bs.length;
  }
  throw new HostError("sorted: unorderable mixed types");
}
function pySorted(arr) {
  return [...arr].sort(pyCompare);
}
function pyMax(arr) {
  if (arr.length === 0) throw new HostError("max: empty sequence");
  return arr.reduce((m, x) => (pyCompare(x, m) > 0 ? x : m));
}
function pyMin(arr) {
  if (arr.length === 0) throw new HostError("min: empty sequence");
  return arr.reduce((m, x) => (pyCompare(x, m) < 0 ? x : m));
}

// The host-independent foreign targets the corpus resolves. The ambient
// os.getcwd is env-specific (process.cwd under Node, a VFS root in a browser),
// so each host supplies it; everything else is shared.
export function sharedTargets(getcwd) {
  return {
    "builtins.sorted": (arr) => pySorted(arr),
    "builtins.max": (arr) => (Array.isArray(arr) ? pyMax(arr) : arr),
    "builtins.min": (arr) => (Array.isArray(arr) ? pyMin(arr) : arr),
    "builtins.str": (x) => (typeof x === "string" ? x : String(x)),
    "time.time": () => Date.now() / 1000,
    "random.random": () => Math.random(),
    "os.getcwd": getcwd,
  };
}

// resolve() shared by every host: a cache, the target table, then the two
// refusals host.py makes — bad target (no dot) and not-found.
export function resolveWith(table, cache, target) {
  if (target in cache) return cache[target];
  if (target in table) {
    cache[target] = table[target];
    return table[target];
  }
  if (!target.includes(".")) throw new HostError(`bad target: ${target}`);
  throw new HostError(`cannot find '${target}'`);
}

// A host with the outside world replaced — an in-memory VFS for files, a
// responses map for ask, and Math.random/Date.now (or a fixed clock) — the
// browser-safe base for both TestHost (hermetic tests) and BrowserHost (the
// browser backend). The JS analogue of host.py's TestHost, minus any Python
// runtime.
export class MemoryHost extends Host {
  constructor({ responses = {}, files = {}, now = null, cwd = "/" } = {}) {
    super();
    this.responses = responses;
    this.files = { ...files };
    this.now = now;
    this.cwd = cwd;
    this.shown = [];
    this.recorded = [];
    this._resolved = {};
    this._targets = sharedTargets(() => this.cwd);
  }
  get name() {
    return "memory";
  }
  ask(url) {
    const r = this.responses;
    if (typeof r === "function") return r(url);
    if (Object.prototype.hasOwnProperty.call(r, url)) return r[url];
    throw new HostError(`no stubbed response for ${url}`);
  }
  read(path) {
    if (!Object.prototype.hasOwnProperty.call(this.files, path)) {
      throw new HostError(`no such file: ${path}`);
    }
    return this.files[path];
  }
  write(path, text) {
    this.files[path] = text;
  }
  show(text) {
    this.shown.push(text);
  }
  clock() {
    return this.now !== null ? this.now : Date.now() / 1000;
  }
  record(entry) {
    this.recorded.push(entry);
  }
  resolve(target) {
    return resolveWith(this._targets, this._resolved, target);
  }
  targetHint() {
    return "`module.function`, e.g. `builtins.sorted`";
  }
  parseJson(text) {
    return JSON.parse(text);
  }
}

// The hermetic host the agreement suite uses — a MemoryHost with a fixed clock,
// so a clock-reading program is reproducible. host.py's TestHost by another
// route (in-memory, not a Python-runtime subclass).
export class TestHost extends MemoryHost {
  constructor({ responses = {}, files = {}, now = 1000000.0 } = {}) {
    super({ responses, files, now });
  }
  get name() {
    return "test";
  }
}
